Assets/Scripts/Framework/Train/TrainController.cs

```diff
@@ -0,0 +1,271 @@
+using UnityEngine.Splines;
+using FishNet.Object;
+using UnityEngine;
+using FishNet;
+using System;
+
+namespace DerailedDeliveries.Framework.Train
+{
+    /// <summary>
+    /// Class responsible for moving train along rails spline.
+    /// </summary>
+    [RequireComponent(typeof(TrainEngine))]
+    public class TrainController : NetworkBehaviour
+    {
+        /// <summary>
+        /// Reference to the spline line data.
+        /// </summary>
+        [field: SerializeField]
+        public SplineContainer Spline { get; private set; }
+
+        [Header("Train Config")]
+        [SerializeField, Range(0, 1)]
+        private float _trainFrontStartTime;
+
+        [SerializeField]
+        private float _heightOffset;
+
+        [Header("Wagons Config")]
+        [SerializeField]
+        private Transform _frontWagon;
+
+        [SerializeField]
+        private float _wagonFollowDistance;
+
+        [SerializeField]
+        private float _wagonSpacing;
+
+        [SerializeField]
+        private Transform[] _wagons;
```

Comment:
Maybe clearify in the naming what wagons are there;
```suggestion
        private Transform[] _allWagons;
```
Or perhaps
```suggestion
        private Transform[] _followingWagons;
```
Mainly confuses me because you have a reference to the front wagon so naming this a bit differently could make the difference.

Reply:
Agreed, I will change it to `_followingWagons`

Reply:
I'm okay with that.